src/viam/sdk/robot/client.cpp

```diff
@@ -57,6 +47,110 @@ using viam::robot::v1::Status;
 // NOLINTNEXTLINE
 const std::string kStreamRemoved("Stream removed");
 
+viam::robot::v1::DiscoveryQuery discovery_query::to_proto() const {
```

Comment:
@stuqdog - I'm considering whether we want some sort of traits class for proto to type interconversion. Think From/Into sort of thing. It'd let us use a uniform syntax for moving between our vocabulary types and the backing protos. I don't have syntax in mind yet, but any thoughts?

Reply:
I like this. I don't think it's urgent but programmatically enforcing a consistent syntax and having a way of clearly expressing in code which types we expect to have proto conversions seems like a win. I'm not positive I have the look of it precisely right in my head, something like the following?
```
class ProtoConvertible {
  public:
   template <typename T, typename ProtoT>
   struct traits;

   template <typename T, typename ProtoT>
   static T from_proto(const ProtoT& proto) {
     return traits<T, ProtoT>::from_proto(proto);
   }

   template <typename T, typename ProtoT>
   static ProtoT to_proto(const T& t) const {
    return traits<T, ProtoT>::to_proto(t);
  }
...
}
```

and then define in `robot_client.hpp` 
``` 
template <>
struct ProtoConvertible::traits<frame_system_config, FrameSystemConfig> {
  static FrameSystemConfig to_proto(const frame_system_config& fsc);
  static frame_system_config to_proto(const FrameSystemConfig& proto);
}
class RobotClient {
  ...
  struct frame_system_config {
   ...
   FrameSystemConfig to_proto() const {
    return ProtoConvertible::to_proto<frame_system_config, FrameSystemConfig>(this);
   } // actual implementation should go in .cpp

   static frame_system_config from_proto(const FrameSystemConfig& proto) {
    return ProtoConvertible::from_proto<frame_system_config, FrameSystemConfig>(proto);
   } // actual implementation should go in .cpp
  }
...
}
```

And then implement all the actual logic in (for this example) `robot_client.cpp`?

Assuming I've got the gist of it right here, it seems like a bit of boilerplate but probably worthwhile for clarity and consistency.

Reply:
Yes, I think that's along the lines of what I had in mind. I filed a ticket so we don't forget - it definitely isn't work to be done in this review.